sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java

```diff
@@ -195,86 +198,86 @@ private Context startTracingSpan(Method method, Context context) {
+                if (hostPath == null || hostPath.isEmpty() || "/".equals(hostPath) || path.contains("://")) {
                     urlBuilder.setPath(path);
                 } else {
                     urlBuilder.setPath(hostPath + "/" + path);
                 }
             }
         }
 
-        for (final EncodedParameter queryParameter : methodParser.setEncodedQueryParameters(args)) {
-            urlBuilder.setQueryParameter(queryParameter.getName(), queryParameter.getEncodedValue());
-        }
+        methodParser.setEncodedQueryParameters(args, urlBuilder);
 
         final URL url = urlBuilder.toUrl();
         final HttpRequest request = configRequest(new HttpRequest(methodParser.getHttpMethod(), url),
             methodParser, args);
 
         // Headers from Swagger method arguments always take precedence over inferred headers from body types
-        for (final HttpHeader header : methodParser.setHeaders(args)) {
-            request.setHeader(header.getName(), header.getValue());
-        }
+        HttpHeaders httpHeaders = request.getHeaders();
+        methodParser.setHeaders(args, httpHeaders);
 
         return request;
     }
 
     @SuppressWarnings("unchecked")
-    private HttpRequest configRequest(HttpRequest request, SwaggerMethodParser methodParser, Object[] args)
-        throws IOException {
+    private HttpRequest configRequest(final HttpRequest request,
+                                      final SwaggerMethodParser methodParser,
+                                      final Object[] args) throws IOException {
         final Object bodyContentObject = methodParser.setBody(args);
         if (bodyContentObject == null) {
             request.getHeaders().put("Content-Length", "0");
         } else {
+            // We read the content type from the @BodyParam annotation
             String contentType = methodParser.getBodyContentType();
+
+            // If this is null or empty, the service interface definition is incomplete and should
+            // be fixed to ensure correct definitions are applied
             if (contentType == null || contentType.isEmpty()) {
                 if (bodyContentObject instanceof byte[] || bodyContentObject instanceof String) {
                     contentType = ContentType.APPLICATION_OCTET_STREAM;
                 } else {
                     contentType = ContentType.APPLICATION_JSON;
                 }
+//                throw logger.logExceptionAsError(new IllegalStateException(
+//                    "The method " + methodParser.getFullyQualifiedMethodName() + " does does not have its content "
+//                        + "type correctly specified in its service interface"));
             }
 
             request.getHeaders().put("Content-Type", contentType);
 
+            // TODO(jogiles) this feels hacky
             boolean isJson = false;
             final String[] contentTypeParts = contentType.split(";");
```

Comment:
`String.split(.)` internally uses java regex which is CPU expensive.

Please ensure regex is not used:
https://www.baeldung.com/java-string-performance#split

Reply:
I looked into this when I was reviewing the code, String.split has a fast-path for single-char 'regexes' which this goes through. You can see the fast path in the JDK here: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/String.java#l2317

Reply:
Thank you @JonathanGiles I learn something today.